Bridging services

haueck opened the pull request:
https://github.com/ros2/ros1_bridge/issues/33

dirk-thomas commented:
I am sorry that my latest PR #44 made this conflict. I will rebase your patch against master and comment here when it is ready.

haueck commented on doc/index.rst:
```diff
@@ -45,37 +49,64 @@ Each mapping rule can have one of three types:
 Example mapping rules file
 --------------------------
 
-Example package mapping rule::
```
Fixed

haueck commented:
Thanks for the review!

Ad. 1. Fixed, factory is not throwing anymore as it is a normal situation when there is no bridge for a service

Ad. 2. This client written in Python makes two calls [add_two_ints_client](https://github.com/ros/ros_tutorials/blob/hydro-devel/rospy_tutorials/005_add_two_ints/add_two_ints_client)

Ad. 3. I added error handling when removing a bridge but I cannot verify it because I cannot build ROS2 with RTI Connext DDS. What Connext version do you use? I use 5.2.3 for open source projects and it fails during linking. When I used a trial version it was fine but it has expired.

Ad. 4. It is probably related to this issue. If the bridge finds a ROS1 server it creates a ROS2 server. If ROS1 server is removed, the bridge removes its own ROS2 server. But this server does not disappear and the bridge thinks that it found a new ROS2 server... so it prepares for routing from ROS1 and ROS2 and will never route again from ROS2 to ROS1.

dirk-thomas commented:
> Ad. 3. I added error handling when removing a bridge but I cannot verify it because I cannot build ROS2 with RTI Connext DDS. What Connext version do you use? I use 5.2.3 for open source projects and it fails during linking. When I used a trial version it was fine but it has expired.

We can build ROS 2 against the professional version (which requires a license) as well as against the evaluation version (which can be downloaded from their website). Maybe you can post more information what exactly fails for you in order to resolve the problem.

haueck commented:
My branch is rebased + one more commit with some updates after rebase.

haueck commented:
Here is a log file from building ROS2 with RTI Connext DDS for open source projects: [gist/rti_connext_dds.log](https://gist.github.com/haueck/ddc9094704e97593513dbea1e3489fd9)

dirk-thomas commented on README.md:
```diff
@@ -274,3 +273,54 @@ rostopic pub -r 1 /flip_image std_msgs/Bool "{data: false}"
+
+While building, ros1_bridge looks for all installed ROS and ROS2 services.
+Found services are matched by comparing package name, service name and fields
+in a request and a response. If all names are the same in ROS and ROS2 service,
```
Please add a newline after every sentence.

haueck replied:
Done

dirk-thomas commented on include/ros1_bridge/factory.hpp:
```diff
@@ -152,6 +152,92 @@ class Factory : public FactoryInterface
     ROS1_T & ros1_msg);
 };
 
+template <class R1_TYPE, class R1_REQ, class R1_RES, class R2_TYPE, class R2_REQ, class R2_RES>
```
Please remove the template arguments for request / response. They can be derived with e.g. `R1_TYPE::Request`.

dirk-thomas replied:
Please use the same template names as for pub / sub: e.g. `ROS1_T`.

haueck replied:
Done

dirk-thomas commented on include/ros1_bridge/factory.hpp:
```diff
@@ -152,6 +152,92 @@ class Factory : public FactoryInterface
+private:
+
+  void translate_1_to_2(const R1_REQ&, R2_REQ&);
+  void translate_1_to_2(const R1_RES&, R2_RES&);
```
These should be able to reuse the conversion functions already being generated for messages. The request and response are normal messages - the only difference is that they are in the `srv` namespace rather then `msg`.

haueck replied:
They do reuse those generated functions. They call convert*() functions for all non-primitive data types.

If you mean, that we could use that template generating convert*() to generate functions converting requests and responses, it is possible but not that trivial. I would need to rewrite this python script and I do not have time to do it now.

dirk-thomas replied:
Yes, currently the implementation of these functions iterate over the fields. I think we can get rid of this of we expand the already existing message templates for the request / response. I will look into it...

dirk-thomas commented on include/ros1_bridge/factory.hpp:
```diff
@@ -152,6 +152,92 @@ class Factory : public FactoryInterface
+  void translate_1_to_2(const R1_REQ&, R2_REQ&);
+  void translate_1_to_2(const R1_RES&, R2_RES&);
+  void translate_2_to_1(R1_REQ&, const R2_REQ&);
+  void translate_2_to_1(R1_RES&, const R2_RES&);
```
Please swap the argument order to match the name of the functions. Might become obsolete with the comment above.

haueck replied:
I swapped the arguments.

dirk-thomas commented on include/ros1_bridge/factory.hpp:
```diff
@@ -152,6 +152,92 @@ class Factory : public FactoryInterface
+    R1_TYPE srv;
+    translate_2_to_1(srv.request, *request);
+    if (client.call(srv))
+    {
```
Please place the `{` and `}` within functions not on a separate line to match the code style of the rest of the file.

haueck replied:
Done

dirk-thomas commented on include/ros1_bridge/factory.hpp:
```diff
@@ -152,6 +152,92 @@ class Factory : public FactoryInterface
+    ServiceBridge1to2 bridge;
+    bridge.client = ros2_node->create_client<R2_TYPE>(name);
+    auto m = &ServiceFactory<R1_TYPE,R1_REQ,R1_RES,R2_TYPE,R2_REQ,R2_RES>::forward_1_to_2;
+    auto f = boost::bind(m, this, bridge.client, _1, _2);
```
Please use `std` instead of `boost`.

haueck replied:
Done

dirk-thomas commented on package.xml:
```diff
@@ -15,10 +15,15 @@
   <build_depend>rclcpp</build_depend>
   <build_depend>rmw_implementation_cmake</build_depend>
   <build_depend>std_msgs</build_depend>
+  <build_depend>diagnostic_msgs</build_depend>
```
This should only be a test dependency.

haueck replied:
Done

dirk-thomas commented on package.xml:
```diff
@@ -15,10 +15,15 @@
   <exec_depend>rclcpp</exec_depend>
   <exec_depend>std_msgs</exec_depend>
 
+  <test_depend>ament_cmake_nose</test_depend>
```
This is already covered by the build dependency on `ament_cmake`.

haueck replied:
Done

haueck replied:
It is required

dirk-thomas commented on resource/get_factory.cpp.em:
```diff
@@ -45,4 +45,23 @@ get_factory(const std::string & ros1_type_name, const std::string & ros2_type_na
   throw std::runtime_error("No template specialization for the pair");
 }
 
+std::unique_ptr<ServiceFactoryInterface> get_service_factory(std::string ros, std::string package, std::string name)
```
Pass strings as `const std::string &`.

Also please use better variable names. What does `ros` mean? `package` -> `package_name`, `name` -> `service_name`.

haueck replied:
Done

dirk-thomas commented on resource/pkg_factories.cpp.em:
```diff
@@ -202,4 +216,90 @@ Factory<
+    @(field["ros1"]["name"])1_it != req1.@(field["ros1"]["name"]).end() &&
+    @(field["ros2"]["name"])2_it != req2.@(field["ros2"]["name"]).end()
+  ) {
+    auto& @(field["ros1"]["name"])1 = *(@(field["ros1"]["name"])1_it++);
```
Please add a space between `auto` and `&`.

dirk-thomas commented on ros1_bridge/__init__.py:
```diff
@@ -68,6 +68,40 @@
 def generate_cpp(output_path, template_dir):
+    data = generate_messages()
+    data.update(generate_services())
+    unique_package_names = set(data["ros2_package_names_msg"] + data["ros2_package_names_srv"])
```
Please use single quotes consistently throughout the file.

dirk-thomas commented on ros1_bridge/__init__.py:
```diff
@@ -68,6 +68,40 @@
+                        m.ros2_msg for m in data["mappings"]
+                        if m.ros2_msg.package_name == ros2_package_name],
+                })
+            print(data_pkg)
```
This debug output should be removed.

dirk-thomas commented on ros1_bridge/__init__.py:
```diff
@@ -175,6 +199,43 @@ def get_ros2_messages():
+        for export in pkg.exports:
+            if export.tagname != 'ros1_bridge':
+                continue
+            if 'service_mapping_rules' not in export.attributes:
```
Please avoid a second different export for these mappings. They should be specified in a single mapping file together with the message mapping.

haueck replied:
Done

dirk-thomas commented on ros1_bridge/__init__.py:
```diff
@@ -213,13 +274,21 @@ def __eq__(self, other):
+        'ros1_service_name',
         'ros2_package_name',
         'ros2_message_name',
+        'ros2_service_name',
```
A single mapping shouldn't be able to have a message name as well as a service name.

haueck replied:
I changed conditionals so it is either message mapping or service mapping, never both at the same time.

haueck replied:
After all, I created two separate classes: MessageMappingRule and ServiceMappingRule.

dirk-thomas commented on ros1_bridge/__init__.py:
```diff
@@ -323,6 +402,79 @@ def determine_message_pairs(ros1_msgs, ros2_msgs, package_pairs, mapping_rules):
+                        "type": ros2_type.rstrip("[]"),
+                        "name": ros2_name,
+                    }
+                })
```
This custom code shouldn't be necessary. Messages can already be converted between ROS 1 and ROS 2. There shouldn't be a need to reimplement this.

haueck replied:
As I mentioned before, it is possible to unify some functions in this script but it might not be that easy and I do not have time to do it now.

dirk-thomas commented on src/dynamic_bridge.cpp:
```diff
@@ -196,6 +206,147 @@ void update_bridge(
+      if (factory) {
+        try {
+          service_bridges_2_to_1[name] = factory->service_bridge_2_to_1(ros1_node, ros2_node, name);
+          // printf("Created 2 to 1 bridge for service %s\n", name.data());
```
These messages should be printed (same as for pub / sub) so that the user sees what is happening.

haueck replied:
Done

dirk-thomas commented on test/test_ros1_bridge.py.in:
```diff
@@ -0,0 +1,213 @@
+    self.assertEqual(None, self.bridge.poll(), "Bridge exited prematurely")
+
+if __name__ == '__main__':
+    unittest.main()
```
Please use 4 space indentation for Python code (see PEP8).

haueck replied:
Done

dirk-thomas commented:
Building the package also results in compiler warning about unused variables (`start1`, `start2`, `goal1`, `goal2`).

dirk-thomas commented on README.md:
```diff
@@ -274,3 +273,54 @@ rostopic pub -r 1 /flip_image std_msgs/Bool "{data: false}"
+# Shell C:
+. <ros-install-dir>/setup.bash
+export ROS_MASTER_URI=http://localhost:11311
+<ros-install-dir>/lib/roscpp_tutorials/add_two_ints_server
```
This line should use `rosrun` to start the node.

haueck replied:
Done

dirk-thomas commented:
It works for me when starting a ROS 1 server and a ROS 2 client. But when I stop the ROS 1 server and start it again, the ROS 2 client doesn't find the server anymore.

dirk-thomas commented:
I committed a fix for the merge conflict.

dirk-thomas commented:
@haueck Please let me know when you will have time to address the comments. Depending on the timeline we might merge the patch as is in order to get it in before the beta freeze and the comments need to be addressed in a follow up PR.

dirk-thomas commented:
I will merge the current state for now. Please create a follow up PR soon.

haueck commented:
@dirk-thomas, thank you for the review!

> Building the package also results in compiler warning about unused variables (start1, start2, goal1, goal2).

I didn't use variables with those names, no idea what it is.

>It works for me when starting a ROS 1 server and a ROS 2 client. But when I stop the ROS 1 server and start it again, the ROS 2 client doesn't find the server anymore.

@dhood already asked about it:

>Ad. 4. It is probably related to this issue. If the bridge finds a ROS1 server it creates a ROS2 server. If ROS1 server is removed, the bridge removes its own ROS2 server. But this server does not disappear and the bridge thinks that it found a new ROS2 server... so it prepares for routing from ROS1 and ROS2 and will never route again from ROS2 to ROS1.

I think I will finish rework tomorrow.

dirk-thomas commented:
Thank you for the quick update!

dhood commented:
FYI the unused variable warnings referred to come from the generated files e.g.:

```
/home/dhood/ros2_ws/build_isolated/ros1_bridge/generated/nav_msgs_factories.cpp: In member function ‘void ros1_bridge::ServiceFactory<R1_TYPE, R1_REQ, R1_RES, R2_TYPE, R2_REQ, R2_RES>::translate_2_to_1(R1_REQ&, const R2_REQ&) [with R1_TYPE = nav_msgs::GetPlan; R1_REQ = nav_msgs::GetPlanRequest_<std::allocator<void> >; R1_RES = nav_msgs::GetPlanResponse_<std::allocator<void> >; R2_TYPE = nav_msgs::srv::GetPlan; R2_REQ = nav_msgs::srv::GetPlan_Request_<std::allocator<void> >; R2_RES = nav_msgs::srv::GetPlan_Response_<std::allocator<void> >]’:
/home/dhood/ros2_ws/build_isolated/ros1_bridge/generated/nav_msgs_factories.cpp:589:9: warning: unused variable ‘start1’ [-Wunused-variable]
   auto& start1 = req1.start;
```

haueck commented:
New pull requests: https://github.com/ros2/ros1_bridge/pull/48 https://github.com/ros2/examples/pull/153

I couldn't run tests though, whatever I launch I get this error:

```
./talker
terminate called after throwing an instance of 'std::runtime_error'
  what():  could not create publisher: type support not from this implementation,
 at /ros2_ws/src/ros2/rmw_fastrtps/rmw_fastrtps_cpp/src/functions.cpp:782,
 at /ros2_ws/src/ros2/rcl/rcl/src/rcl/publisher.c:81
Aborted (core dumped)
```

I use current master branches: https://raw.githubusercontent.com/ros2/ros2/master/ros2.repos

haueck commented:
New pull requests: https://github.com/ros2/ros1_bridge/pull/48 https://github.com/ros2/examples/pull/153

Karsten1987 commented:
The PR has leads to a crash of the bridge whenever any topic happens to be of length 4.
In particular, it's this line: https://github.com/ros2/ros1_bridge/blob/master/src/dynamic_bridge.cpp#L651

Consider the following example:

```
std::string first = "scan";
std::string second = "sensor_msgs/LaserScan";
std::string suffix = "reply";
	
size_t suffix_found = first.rfind(suffix);  // return std::string::npos (-1)
printf("%u\n", suffix_found);  // size_t (-1) = 4294967295
printf("%u\n", first.size()); // 4
printf("%u\n", suffix.size()); // 5
printf("%u\n", first.size() - suffix.size());  // size_t(-1) = 4294967295
	
// This expression will evaluate to true, even though the suffix was never found.
if (first.rfind(suffix) == (first.size() - suffix.size())) {
    std::string & t = second;
    // CRASH HERE!
    std::string name(first.begin(), first.end() - suffix.size()); 
} else {
  ...
}
```

One possible solution is to explicitly filter out `std::string::npos` in the case the suffix `Reply` is not found.

haueck commented:
@Karsten1987, thank you for this information! I've created a fix for the issue https://github.com/ros2/ros1_bridge/pull/58
